pkg/restore/restore.go

```diff
@@ -373,103 +381,104 @@ func (ctx *restoreContext) execute() (Result, Result) {
+		ticker := time.NewTicker(1 * time.Second)
+		var lastUpdate *progressUpdate
+		for {
+			select {
+			case <-quit:
+				ticker.Stop()
+				return
+			case val := <-update:
+				lastUpdate = &val
+			case <-ticker.C:
+				if lastUpdate != nil {
+					patch := fmt.Sprintf(`{"status":{"progress":{"totalItems":%d,"itemsRestored":%d}}}`, lastUpdate.totalItems, lastUpdate.itemsRestored)
+					if _, err := ctx.restoreClient.Restores(ctx.restore.Namespace).Patch(context.TODO(), ctx.restore.Name, types.MergePatchType, []byte(patch), metav1.PatchOptions{}); err != nil {
+						ctx.log.WithError(errors.WithStack((err))).Warn("Got error trying to update restore's status.progress")
+					}
+					lastUpdate = nil
+				}
 			}
+		}
+	}()
 
-			// if we don't know whether this namespace exists yet, attempt to create
-			// it in order to ensure it exists. Try to get it from the backup tarball
-			// (in order to get any backed-up metadata), but if we don't find it there,
-			// create a blank one.
-			if namespace != "" && !existingNamespaces.Has(targetNamespace) {
-				logger := ctx.log.WithField("namespace", namespace)
-				ns := getNamespace(logger, archive.GetItemFilePath(ctx.restoreDir, "namespaces", "", namespace), targetNamespace)
-				if _, err := kube.EnsureNamespaceExistsAndIsReady(ns, ctx.namespaceClient, ctx.resourceTerminatingTimeout); err != nil {
-					errs.AddVeleroError(err)
+	// i: iteration counter, totalItems: previously discovered items,
+	totalItems, i, existingNamespaces := 0, 0, sets.NewString()
+
+	for _, selectedResource := range selectedResourceCollection {
+		totalItems += selectedResource.totalItems
+	}
+	for _, selectedResource := range selectedResourceCollection {
+		groupResource := schema.ParseGroupResource(selectedResource.resource)
+
+		for namespace, selectedItems := range selectedResource.selectedItemsByNamespace {
+			for _, selectedItem := range selectedItems {
+				// if we don't know whether this namespace exists yet, attempt to create
+				// it in order to ensure it exists. Try to get it from the backup tarball
+				// (in order to get any backed-up metadata), but if we don't find it there,
+				// create a blank one.
+				if namespace != "" && !existingNamespaces.Has(selectedItem.targetNamespace) {
+					logger := ctx.log.WithField("namespace", namespace)
+					ns := getNamespace(logger, archive.GetItemFilePath(ctx.restoreDir, "namespaces", "", namespace), selectedItem.targetNamespace)
+					if _, nsCreated, err := kube.EnsureNamespaceExistsAndIsReady(ns, ctx.namespaceClient, ctx.resourceTerminatingTimeout); err != nil {
+						errs.AddVeleroError(err)
+						continue
+					} else {
+						// add the newly created namespace to the list of restored items
+						if nsCreated {
+							itemKey := velero.ResourceIdentifier{
+								GroupResource: kuberesource.Namespaces,
+								Namespace:     ns.Namespace,
+								Name:          ns.Name,
+							}
+							ctx.restoredItems[itemKey] = struct{}{}
+						}
+					}
```

Comment:
This block code looks similar to https://github.com/vmware-tanzu/velero/blob/b527a8a32deb9f5546327314b8347495f7855dc5/pkg/restore/restore.go#L794-L811

And the code logic would run twice. I'm not sure what's the difference between these two or could we merge them into one?

Reply:
@jenting This duplication is required because `restoreItem()` can be called for additional items returned by plugins. The recursive call in that case only triggers restoreItem() function alone and not the rest of the previous logic.